stan/math/prim/mat/fun/append_array.hpp

```diff
@@ -1,80 +1,141 @@
-#ifndef STAN_MATH_PRIM_ARR_FUN_APPEND_ARRAY_HPP
-#define STAN_MATH_PRIM_ARR_FUN_APPEND_ARRAY_HPP
+#ifndef STAN_MATH_PRIM_MAT_FUN_APPEND_ARRAY_HPP
+#define STAN_MATH_PRIM_MAT_FUN_APPEND_ARRAY_HPP
 
 #include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/prim/scal/meta/return_type.hpp>
+#include <stan/math/prim/mat/meta/append_return_type.hpp>
+#include <stan/math/prim/scal/err/check_consistent_sizes.hpp>
 #include <vector>
 
 namespace stan {
   namespace math {
+    /**
+     * This function copies the value at x to y, casting from the type of T1 to
```

Comment:
I'm not going to double-review this, but I will pack all my notes into a review so as not to spam everyone on our lists.  

I think this function just needs to be renamed to something like `convert_copy` --- I'm guessing @mitzimorris was thinking this was actually appending arrays.